### PR TITLE
[WIP] Enable WML Training

### PR DESCRIPTION
**Pull request by splovyt**
This PR enables WML training functionality.

- Restructured the original assets and uploaded them to `max-assets-prod`
- Updated the Dockerfile for training
- Bump the API version to `2.0.0`
- Added `Train` section to main README
- Added the cookiecutter training and setup scripts
- Added the model-specific training scripts

TODO:
- [x] Optimize code
- [x] Update data `README.md`
- [x] Update training `README.md`

### CODE REVIEW

**Comment by splovyt** [comment]
@ptitzler @SSaishruthi This PR is ready for review now.

**Comment by splovyt** [comment]
This build also succeeds with my jenkins pipeline :-)

File: training/README.md [hunk 1]
```diff
@@ -0,0 +1,160 @@
+## Train the Model with Your Own Data
+
+This document provides instructions to train the model on Watson Machine Learning, an offering of IBM Cloud. The instructions in this document assume that you already have an IBM Cloud account. If not, please create an [IBM Cloud](https://ibm.biz/Bdz2XM) account. 
+
+- [Prepare Data for Training](#prepare-data-for-training)
+- [Train the Model](#train-the-model)
+- [Rebuild the Model Serving Microservice](#rebuild-the-model-serving-microservice)
```
Comment (ptitzler) [inline training/README.md:7]: model-serving (link text should match section heading)

File: training/README.md [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+### Run the Setup Script
+
+The `wml_setup.py` script prepares your local environment and your IBM Cloud resources for model training.
+
+1. Locate the training configuration file. It is named `max-text-classifier-training-config.yaml`.
+
+   ```
+
+   $ ls *.yaml
+     <max-text-classifier-training-config.yaml>
```
Comment (ptitzler) [inline training/README.md]: `max-text-classifier-training-config.yaml` (no <>)

File: training/data_preparation/README.md [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+## Prerequisites
+No special prerequisites other than the standard data ETL tools to create a `.tsv`-formatted file.
+
+## Preparing your data
+The training data must be a sole `.tsv`-formatted file and must carry the `.tsv` extension. 
+Multiple `.tsv` files in the input bucket for training will result in an error.
+
+The file has to be made up of two columns exactly, separated by the tab-delimiter (`\t`) as is typical for a `tsv` file.
+The entry before the `tab`-delimiter must equal a piece of text, and the element after the delimiter must be the class label corresponding to the text before the delimiter. No headers are accepted.
```
Comment (ptitzler) [inline training/data_preparation/README.md]: We need to be more specific. It's not clear what class labels are acceptable. The sample implies that supported labels are `pos` and `neg`. Is `positive` a valid value? The model description indicates that the model is used to measure positive and negative sentiment. What if somebody uses a class label such as `neutral`?

**Review by ptitzler — DISMISSED** [review]
Still testing.

File: README.md [hunk 4]
```diff
@@ -6,7 +6,7 @@
 
 This repository contains code to instantiate and deploy a text sentiment classifier. This model is able to detect whether a text fragment leans towards a positive or a negative sentiment. Optimal input examples for this model are short strings (preferably a single sentence) with correct grammar, although not a requirement.
 
-The model is based on the [pre-trained BERT-Base, English Uncased](https://github.com/google-research/bert/blob/master/README.md) model and was finetuned on the [IBM Claim Stance Dataset](http://www.research.ibm.com/haifa/dept/vst/debating_data.shtml). The model files are hosted on
+The model is based on the [pre-trained BERT-Base, English Uncased](https://github.com/google-research/bert/blob/master/README.md) model and was fine-tuned on the [IBM Claim Stance Dataset](http://www.research.ibm.com/haifa/dept/vst/debating_data.shtml). The model files are hosted on
 [IBM Cloud Object Storage](http://s3.us-south.cloud-object-storage.appdomain.cloud/max-assets-prod/max-text-sentiment-classifier/1.0.0/assets.tar.gz).
```
Comment (MLnick) [inline README.md]: This COS link also needs to be updated to version `1.2.0` right?

File: tests/test.py [hunk 5]
```diff
@@ -63,9 +63,9 @@ def test_response():
 
     # verify that 'good string' is in fact positive
     print(response['predictions'][0])
```
Comment (MLnick) [inline tests/test.py]: what's the `print` here for?

File: tests/training/test_sample_training.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
+#
```
Comment (MLnick) [inline tests/training/test_sample_training.py]: This seems like we are re-creating tests from the deployable model? Ideally we don't want to duplicate tests.

File: training/README.md (hunk 1 above)
Comment (MLnick) [reply training/README.md:7]: Should we not just say "Rebuild the Model Microservice"? (this is a general comment not just for this model)

File: config.py [hunk 7]
```diff
@@ -25,11 +25,11 @@
...
-API_VERSION = '1.1.0'
+API_VERSION = '2.0.0'
 
 # default model
 MODEL_NAME = 'sentiment_BERT_base_uncased'
-DEFAULT_MODEL_PATH = 'assets/{}'.format(MODEL_NAME)
+DEFAULT_MODEL_PATH = 'assets'
 
 # the metadata of the model
 MODEL_META_DATA = {
```
Comment (MLnick) [inline config.py:34]: model name here should also be MAX Text Sentiment Classifier (same as `API_TITLE`)

File: config.py [hunk 8]
```diff
@@ -25,11 +25,11 @@
 API_TITLE = 'MAX Text Sentiment Classifier'
 API_DESC = 'Detect the sentiment captured in short pieces of text. ' \
            'The model was finetuned on the IBM Project Debater Claim Sentiment dataset.'
-API_VERSION = '1.1.0'
+API_VERSION = '2.0.0'
 
 # default model
 MODEL_NAME = 'sentiment_BERT_base_uncased'
```
Comment (MLnick) [inline config.py]: As such this is probably no longer required?

File: training/data_preparation/README.md (hunk 3 above)
Comment (MLnick) [reply training/data_preparation/README.md]: I believe the labels can be arbitrary (either binary labels like "positive" or "negative" or more than two labels) -  is that correct @splovyt?

File: training/data_preparation/README.md [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+- [Preparing your data](#preparing-your-data)
+
+## Prerequisites
+No special prerequisites other than the standard data ETL tools to create a `.tsv`-formatted file.
+
+## Preparing your data
+The training data must be a sole `.tsv`-formatted file and must carry the `.tsv` extension. 
+Multiple `.tsv` files in the input bucket for training will result in an error.
+
+The file has to be made up of two columns exactly, separated by the tab-delimiter (`\t`) as is typical for a `tsv` file.
```
Comment (MLnick) [inline training/data_preparation/README.md:14]: I think it will be more clear here to provide a table outlining the TSV schema:

| text | label |
| - | - |
| string | string |

Comment (MLnick) [reply training/data_preparation/README.md:14]: Together with an example:

| text | label |
| - | - |
| Exposure to violent video games causes at least a temporary increase in aggression and this exposure correlates with aggression in the real world | neg |
| video game violence is not related to serious aggressive behavior in real life | pos |
| some violent video games may actually have a prosocial effect in some contexts | pos |

File: training/setup/config.json [hunk 10]
```diff
@@ -0,0 +1,11 @@
+{
+  "wml": {
```
Comment (MLnick) [inline training/setup/config.json:2]: Should these be part of this? Seems like credential leaking here @splovyt @ptitzler ?

File: training/training_code/prepare_environment.py [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+    model_file = FLAGS.MODEL_FILE
+    model_folder = FLAGS.MODEL_FOLDER
+
+    archive_path = os.path.join(pretrained_model_dir, model_file)
+    folder_path = os.path.join(pretrained_model_dir, model_folder)
+
+    if not os.path.exists(folder_path):
+        # Downloading the model file (if the unzipped file doesnt exist yet)
+        print('Downloading pre-trained model file...')
+        opener = urllib.request.URLopener()
```
Comment (MLnick) [inline training/training_code/prepare_environment.py:77]: Is WML able to call externally to download?

File: training/training_code/train-max-model.sh [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+
+# Verify whether the DATA_DIR contains only one tsv file, and assign it to the TRAINING_DATA variable.
+tsv_count=`ls ${DATA_DIR}/*.tsv 2>/dev/null | wc -l`
+if [ $tsv_count -gt 1 ]; then
+    echo "ERROR: There are multiple .tsv files present in the data directory."
+    exit $ENV_ERROR_RETURN_CODE
+else
+    # Assign it to the TRAIN_DATA variable
+    TRAINING_DATA=`ls ${DATA_DIR}/*.tsv`
+fi
```
Comment (MLnick) [inline training/training_code/train-max-model.sh:53]: Are we supporting a validation dataset? Or does the `bert` training code take care of that?

File: training/training_code/training_requirements.txt [hunk 13]
```diff
@@ -0,0 +1 @@
+tensorflow == 1.12.2
```
Comment (MLnick) [inline training/training_code/training_requirements.txt]: If running in a WML environment, we actually don't need to pip install the framework as it's already available.

**Review by MLnick — COMMENTED** [review]
First pass review comments

File: training/data_preparation/README.md (hunk 3 above)
Comment (splovyt) [reply training/data_preparation/README.md]: Yes it's single-label, multi-class if that makes sense :). So we can have multiple classes such as `very positive`, `positive`, `neutral`, `negative`, `very negative` if we wanted, but only one of those can be assigned at a time. I will make it clear in the docs!

File: training/setup/config.json (hunk 10 above)
Comment (splovyt) [reply training/setup/config.json:2]: @ptitzler how did this occur?

File: training/training_code/prepare_environment.py (hunk 11 above)
Comment (splovyt) [reply training/training_code/prepare_environment.py:77]: Apparently yes :-)

File: training/README.md (hunk 1 above)
Comment (splovyt) [reply training/README.md:7]: Will make Patrick's change, but will leave the conversation unresolved until get some feedback on this

File: config.py (hunk 8 above)
Comment (splovyt) [reply config.py]: What do you mean?

File: training/data_preparation/README.md (hunk 9 above)
Comment (splovyt) [reply training/data_preparation/README.md:14]: Thank you. I have added the first table for illustration, but the second one I have kept in the format it is now cause I'd like it to be as close to the realistic dataset expectations.

File: tests/test.py (hunk 5 above)
Comment (splovyt) [reply tests/test.py]: Oops!

File: config.py (hunk 7 above)
Comment (splovyt) [reply config.py:34]: why do we have the `MODEL_NAME` variable even if we set it to `API_TITLE` anyways? Should we remove it from the skeleton?

File: config.py (hunk 8 above)
Comment (splovyt) [reply config.py]: Oh I think I understand now, and I agree that we can remove this.

File: config.py (hunk 7 above)
Comment (splovyt) [reply config.py:34]: (see next review comment)

File: tests/training/test_sample_training.py (hunk 6 above)
Comment (splovyt) [reply tests/training/test_sample_training.py]: I'll split up the test files.

File: training/training_code/train-max-model.sh (hunk 12 above)
Comment (splovyt) [reply training/training_code/train-max-model.sh:53]: Currently it doesn't support a validation dataset, but it could be added in the future.

**Comment by splovyt** [comment]
I have made both your changes and both Jenkins and Travis are succeeding.

File: training/training_code/prepare_environment.py (hunk 11 above)
Comment (MLnick) [reply training/training_code/prepare_environment.py:77]: huh interesting...

File: tests/test_api.py [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+def test_metadata():
+
+    model_endpoint = 'http://localhost:5000/model/metadata'
+
+    r = requests.get(url=model_endpoint)
+    assert r.status_code == 200
+
+    metadata = r.json()
+    assert metadata['id'] == 'max-text-sentiment-classifier'
+    assert metadata['name'] == 'Bert Base Uncased TensorFlow Model'
```
Comment (MLnick) [inline tests/test_api.py]: name should be "MAX Text Sentiment Classifier"

File: training/README.md [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+    $ python wml_setup.py max-text-classifier-training-config.yaml 
+     ...
+   ```
+   
+1. After setup has completed, define the displayed environment variables. These variables provide the model training script with access credentials for your Watson Machine Learning service and Cloud Object Storage service. 
+
+   MacOS example:
+
+   ```
+   $ export ML_INSTANCE=...
```
Comment (MLnick) [inline training/README.md:58]: This needs to be brought in line with most recent training framework readme template

File: tests/test_api.py (hunk 14 above)
Comment (MLnick) [reply tests/test_api.py]: config / metadata and test case should be updated

**Comment by MLnick** [comment]
A couple minor comments, otherwise looks ready

File: .gitignore [hunk 16]
```diff
@@ -1,3 +1,5 @@
+training/training_output/
+*-model-building-code.zip
```
Comment (bdwyer2) [inline .gitignore:2]: would you mind adding `.idea/` to `.gitignore` while you're at it?

File: training/README.md [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+     # Checking environment variables ...
+     # --------------------------------------------------------
+     ...
+   ```
+
+   If preparation completed successfully:
+
+    - The required environment variables are defined.
+    - Training data is present in the Cloud Object Storage bucket that Watson Machine Learning will access to train the model.
+    - The model training code is packaged in a ZIP file named ` max-text-classifier-model-building-code.zip` that Watson Machine Learning uses to train the model.
```
Comment (bdwyer2) [inline training/README.md]:
```suggestion
    - The model training code is packaged in a ZIP file named `max-text-classifier-model-building-code.zip` that Watson Machine Learning uses to train the model.
```

File: training/setup/setup_functions.py [hunk 18]
```diff
@@ -0,0 +1,356 @@
...
+                               the key retrieved.
+        :return: intermediate list and count
+        """
+        for index, key in enumerate(available_keys['resources'], start=1):
+            count += 1
+            inter_wml_list.append([count, key['name'], key['guid']])
+
+        return inter_wml_list, count
+
+    def wml_key_check(self, instance_guid): # noqa
```
Comment (bdwyer2) [inline training/setup/setup_functions.py:182]: The `# noqa`'s in this file don't seem necessary

File: tests/test_api.py (hunk 14 above)
Comment (splovyt) [reply tests/test_api.py]: Done, thanks Nick.

File: .gitignore (hunk 16 above)
Comment (splovyt) [reply .gitignore:2]: Of course!

File: training/setup/setup_functions.py (hunk 18 above)
Comment (splovyt) [reply training/setup/setup_functions.py:182]: Thanks for the feedback, but I can't really change it here.. The setup scripts are managed here:
https://github.ibm.com/CODAIT/max-model-training/

Comment (SSaishruthi) [reply training/setup/setup_functions.py:182]: Tracking all changes related to training framework separately. Will make all in one.

Comment (splovyt) [reply training/setup/setup_functions.py:182]: Thanks :)))

File: training/README.md (hunk 15 above)
Comment (splovyt) [reply training/README.md:58]: Thanks Nick 🙂updated this!

**Comment by splovyt** [comment]
Thanks all, have updated this!

**Review by MLnick — DISMISSED** [review]
LGTM

**Comment by MLnick** [comment]
@splovyt branch needs to be updated to merge in latest master

**Review by MLnick — APPROVED** [review]
